core/node/vm_runner/src/storage.rs

```diff
@@ -34,13 +34,7 @@ pub trait StorageLoader: ReadStorageFactory {
     async fn load_batch(
         &self,
         l1_batch_number: L1BatchNumber,
-    ) -> anyhow::Result<Option<BatchExecuteData>>;
-
-    /// A workaround for Rust's limitations on upcasting coercion. See
-    /// https://github.com/rust-lang/rust/issues/65991.
-    ///
-    /// Should always be implementable as [`StorageLoader`] requires [`ReadStorageFactory`].
-    fn upcast(self: Arc<Self>) -> Arc<dyn ReadStorageFactory>;
+    ) -> anyhow::Result<Option<(BatchExecuteData, OwnedStorage)>>;
```

Comment:
With this change, there's no more data race possible between getting `BatchExecuteData` and `BatchExecutor` logic. IMO, it's cleaner as well.